File: batou-rules/testdata/fixtures/java/safe/XssEscaped.java
package com.example.safe;

import java.io.IOException;
import java.io.PrintWriter;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import org.owasp.encoder.Encode;

public class XssEscaped extends HttpServlet {

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        String name = request.getParameter("name");
        String query = request.getParameter("q");

        response.setContentType("text/html; charset=UTF-8");
        PrintWriter out = response.getWriter();
        out.println("<html><body>");
        out.println("<h1>Hello, " + Encode.forHtml(name) + "</h1>");
        out.println("<p>Search results for: " + Encode.forHtml(query) + "</p>");
        out.println("</body></html>");
    }

    @Override
    protected void doPost(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        String feedback = request.getParameter("feedback");
        String safeFeedback = Encode.forHtml(feedback);
        PrintWriter writer = response.getWriter();
        writer.write("<div class='feedback'>" + safeFeedback + "</div>");
    }
}

File: batou-rules/testdata/fixtures/php/safe/xss_escaped.php
<?php
// Safe: All user output properly escaped with htmlspecialchars.
// No XSS possible - all dynamic content is HTML-encoded.

$name = $_GET['name'] ?? '';
$search = $_GET['search'] ?? '';
?>
<!DOCTYPE html>
<html>
<head>
    <title>Greeting Page</title>
    <meta charset="UTF-8">
</head>
<body>

<h1>Hello, <?php echo htmlspecialchars($name, ENT_QUOTES, 'UTF-8'); ?></h1>

<form method="GET" action="">
    <label for="name">Enter your name:</label>
    <input type="text" id="name" name="name" value="<?php echo htmlspecialchars($name, ENT_QUOTES, 'UTF-8'); ?>" />
    <input type="submit" value="Submit" />
</form>

<?php if ($search): ?>
    <p>Search results for: <?php echo htmlspecialchars($search, ENT_QUOTES, 'UTF-8'); ?></p>
<?php endif; ?>

<?php
// Safe: displaying database content with escaping
$conn = new mysqli("localhost", getenv("DB_USER"), getenv("DB_PASS"), "dvwa");
$result = $conn->query("SELECT author, body FROM comments ORDER BY created_at DESC");

while ($row = $result->fetch_assoc()) {
    echo "<div class='comment'>";
    echo "<strong>" . htmlspecialchars($row['author'], ENT_QUOTES, 'UTF-8') . "</strong>";
    echo "<p>" . htmlspecialchars($row['body'], ENT_QUOTES, 'UTF-8') . "</p>";
    echo "</div>";
}

// Safe: using strip_tags for plain text extraction
$bio = strip_tags($_POST['bio'] ?? '');
echo "<p>" . htmlspecialchars($bio, ENT_QUOTES, 'UTF-8') . "</p>";
?>

</body>
</html>
